File: template/sort/sort.go
package sort

func QuickSort(array []int) {
	beg, end := 0, len(array)-1
	qsort(array, beg, end)
}

func xswap(array []int, i, j int) {
	if i != j {
		array[i], array[j] = array[j], array[i]
	}
}

func qsort(array []int, beg, end int) {
	if beg >= end {
		return
	}
	pivot := beg
	for i := beg + 1; i <= end; i++ {
		if array[i] <= array[beg] {
			pivot++
			xswap(array, i, pivot)
		}
	}
	xswap(array, beg, pivot)
	qsort(array, beg, pivot-1)
	qsort(array, pivot+1, end)
}

func MergeSort(array []int) {
	tmp := make([]int, len(array))
	beg, end := 0, len(array)-1
	msort(array, tmp, beg, end)
}

func msort(array, tmp []int, beg, end int) {
	if beg >= end {
		return
	}
	mid := beg + (end-beg)/2
	msort(array, tmp, beg, mid)
	msort(array, tmp, mid+1, end)
	i, j, k := beg, mid+1, beg
	for i <= mid && j <= end {
		if array[i] <= array[j] {
			tmp[k] = array[i]
			i++
			k++
		} else {
			tmp[k] = array[j]
			j++
			k++
		}
	}
	for i <= mid {
		tmp[k] = array[i]
		i++
		k++
	}
	for j <= end {
		tmp[k] = array[j]
		j++
		k++
	}
	for i := beg; i <= end; i++ {
		array[i] = tmp[i]
	}
}

func HeapSort(array []int) {
	for i := len(array) - 1; i >= 0; i-- {
		hdown(array, i, len(array))
	}
	for i := len(array) - 1; i != 0; i-- {
		xswap(array, 0, i)
		hdown(array, 0, i)
	}
}

func hdown(array []int, p int, size int) {
	for p < size {
		l := p*2 + 1
		r := p*2 + 2
		m := p
		if l < size && array[l] > array[m] {
			m = l
		}
		if r < size && array[r] > array[m] {
			m = r
		}
		if p == m {
			return
		}
		xswap(array, p, m)
		p = m
	}
}

File: src/main/java/template/Sort.java
package template;

public class Sort {

    private static void xswap(int[] data, int i, int j) {
        int t = data[i];
        data[i] = data[j];
        data[j] = t;
    }

    private static void msort(int[] data, int[] buff, int beg, int end) {
        if (beg >= end) {
            return;
        }
        int mid = beg + (end - beg) / 2;
        msort(data, buff, beg, mid);
        msort(data, buff, mid + 1, end);
        int i = beg, j = mid + 1;
        int k = beg;
        while (i <= mid && j <= end) {
            if (data[i] <= data[j]) {
                buff[k++] = data[i++];
            } else {
                buff[k++] = data[j++];
            }
        }
        while (i <= mid) {
            buff[k++] = data[i++];
        }
        while (j <= end) {
            buff[k++] = data[j++];
        }
        System.arraycopy(buff, beg, data, beg, end - beg + 1);
    }

    public static void msort(int[] data) {
        if (data.length > 1) {
            int[] buff = new int[data.length];
            msort(data, buff, 0, data.length - 1);
        }
    }

    private static void hdown(int[] data, int i, int n) {
        while (i < n) {
            int l = i * 2 + 1;
            int r = i * 2 + 2;
            int m = i;
            if (l < n && data[l] > data[m]) {
                m = l;
            }
            if (r < n && data[r] > data[m]) {
                m = r;
            }
            if (m == i) {
                return;
            }
            xswap(data, i, m);
            i = m;
        }
    }

    public static void hsort(int[] data) {
        if (data.length > 1) {
            int n = data.length;
            for (int i = n - 1; i >= 0; i--) {
                hdown(data, i, n);
            }
            for (int i = n - 1; i != 0; i--) {
                xswap(data, 0, i);
                hdown(data, 0, i);
            }
        }
    }

    private static void qsort(int[] data, int beg, int end) {
        if (beg >= end) {
            return;
        }
        int mid = beg + (end - beg) / 2;
        xswap(data, beg, mid);
        int i = beg;
        for (int j = i + 1; j <= end; j++) {
            if (data[j] < data[beg]) {
                xswap(data, j, ++i);
            }
        }
        xswap(data, beg, i);
        qsort(data, beg, i - 1);
        qsort(data, i + 1, end);
    }

    public static void qsort(int[] data) {
        if (data.length > 1) {
            qsort(data, 0, data.length - 1);
        }
    }

}
